File: 144 Binary Tree Preorder Traversal/144.binary-tree-preorder-traversal.py
#
# @lc app=leetcode id=144 lang=python3
#
# [144] Binary Tree Preorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
    # Recursive solution
        result = []
        def preorder(node):
            if not node:
                return
            result.append(node.val)
            preorder(node.left)
            preorder(node.right)
        
        preorder(root)
        return result
    
    # Iterative solution
    '''
        if not root:
            return []
        
        result = []
        stack = [root]
        
        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        
        return result
    '''
  
# @lc code=end

File: 144 Binary Tree Preorder Traversal/144.binary-tree-preorder-traversal.cpp
/*
 * @lc app=leetcode id=144 lang=cpp
 *
 * [144] Binary Tree Preorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>
#include <stack> // for iterative solution
using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        preorder(root, result);  // Call the helper function with root and result vector.
        return result;
    }

private:
    void preorder(TreeNode* node, vector<int>& result) {
        // 1. Base case: if node is null, return.
        if (!node) return;
        
        // 2. Visit the node (append its value).
        result.push_back(node->val);
        
        // 3. Traverse the left subtree.
        preorder(node->left, result);
        
        // 4. Traverse the right subtree.
        preorder(node->right, result);
    }
};

/* iterative solution
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        if (!root) return result;  // 1. If the root is null, return an empty result.

        stack<TreeNode*> stack;
        stack.push(root);  // 2. Initialize the stack with the root node.

        while (!stack.empty()) {
            // 3. Pop the top node from the stack and visit it.
            TreeNode* node = stack.top();
            stack.pop();
            result.push_back(node->val);

            // 4. Push the right child first (if exists), so the left child is processed first.
            if (node->right) stack.push(node->right);
            if (node->left) stack.push(node->left);
        }

        return result;  // 5. Return the result containing the preorder traversal.
    }
};
*/

// @lc code=end

